app/models/concerns/with_import_action.rb

```diff
@@ -0,0 +1,60 @@
+module WithImportAction
+    
+  def test_code(root)
+    [file(root, 'test'), extension(root, 'test')]
+  end
+  
+  def extra(dir)
+    file(dir, 'extra')
+  end
+  
+  def description(root)
+    markdown(root, 'description')
+  end
+
+  def hint(root)
+    markdown(root, 'hint')
+  end
+  
+  def meta(root)
+    yaml(root, 'meta')
+  end
+
+  def expectations(root)
+    yaml(root, 'expectations')
+  end
+
+
+  def file(root, filename)
+    search_file(root, filename, lambda {|file| read_file file})
+  end
+  
+  def extension(root,filename)
+    search_file(root, filename, lambda {|file| parse_extension file})
+  end
```

Comment:
The problem of splitting extension and file fetch is that now you are actually performing the same  filesystem operation twice, which is redundant. 

Aside of that, please notice that is you write

```
def search_file(root, filename, &func)
``` 

You can use then this method this way:

```ruby
search_file(root, filename)  {|file| read_file file})
```

Which is cleaner